src/app/qgsgeometryvalidationdock.cpp

```diff
@@ -0,0 +1,343 @@
+  mErrorLocationRubberband->hide();
+  mErrorLocationRubberband->update();
+
+  onCurrentErrorChanged( currentIndex(), QModelIndex() );
+}
+
+QModelIndex QgsGeometryValidationDock::currentIndex() const
+{
+  if ( !mGeometryValidationModel->rowCount() )
+    return QModelIndex();
+  else
+    return mErrorListView->selectionModel()->currentIndex();
+}
+
+void QgsGeometryValidationDock::onCurrentErrorChanged( const QModelIndex &current, const QModelIndex &previous )
+{
+  Q_UNUSED( previous )
+
+  mProblemDetailWidget->setVisible( current.isValid() );
+
+  mNextButton->setEnabled( current.isValid() && current.row() < mGeometryValidationModel->rowCount() - 1 );
+  mPreviousButton->setEnabled( current.isValid() && current.row() > 0 );
+
+  mProblemDetailWidget->setVisible( current.isValid() );
+
+  if ( !current.isValid() )
+    return;
+
+  mProblemDescriptionLabel->setText( current.data( QgsGeometryValidationModel::DetailsRole ).toString() );
+
+  QgsGeometryCheckError *error = current.data( QgsGeometryValidationModel::GeometryCheckErrorRole ).value<QgsGeometryCheckError *>();
+  if ( error )
+  {
+    while ( QWidget *w = mResolutionWidget->findChild<QWidget *>() )
+      delete w;
+
+    delete mResolutionWidget->layout();
+
+    if ( error->status() != QgsGeometryCheckError::StatusFixed )
+    {
+      const QStringList resolutionMethods = error->check()->resolutionMethods();
+      QGridLayout *layout = new QGridLayout( mResolutionWidget );
+      int resolutionIndex = 0;
+      for ( const QString &resolutionMethod : resolutionMethods )
+      {
+        QToolButton *resolveBtn = new QToolButton( mResolutionWidget );
+        resolveBtn->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/algorithms/mAlgorithmCheckGeometry.svg" ) ) );
+        layout->addWidget( resolveBtn, resolutionIndex, 0 );
+        QLabel *resolveLabel = new QLabel( resolutionMethod, mResolutionWidget );
+        resolveLabel->setWordWrap( true );
+        layout->addWidget( resolveLabel, resolutionIndex, 1 );
+        connect( resolveBtn, &QToolButton::clicked, this, [resolutionIndex, error, this]()
+        {
+          mGeometryValidationService->fixError( error, resolutionIndex );
+        } );
+        resolutionIndex++;
+      }
+
+      mResolutionWidget->setLayout( layout );
+
+      showHighlight( current );
+    }
+  }
+
+  bool hasFeature = !FID_IS_NULL( current.data( QgsGeometryValidationModel::ErrorFeatureIdRole ) );
+  mZoomToFeatureButton->setEnabled( hasFeature );
+
+  switch ( mLastZoomToAction )
+  {
+    case  ZoomToProblem:
+      zoomToProblem();
+      break;
+
+    case ZoomToFeature:
+      zoomToFeature();
+      break;
+  }
+}
+
+void QgsGeometryValidationDock::onCurrentLayerChanged( QgsMapLayer *layer )
+{
+  if ( layer == mCurrentLayer )
+    return;
+
+  if ( mCurrentLayer )
+  {
+    disconnect( mCurrentLayer, &QgsVectorLayer::editingStarted, this, &QgsGeometryValidationDock::onLayerEditingStatusChanged );
```

Comment:
I think this is causing a crash when reloading project.
I guess we need to remove the connection when the layer is removed from the layer tree? 
Or can we use some smart pointer here?

Reply:
Do you have a stack trace?

Reply:
Latest commit should have changed that anyway

Reply:
it was on a current layer changed from the layer tree, don't have it anymore.
just out of curiosity, why can't we use a scoped pointer here?

Reply:
A `QPointer` would probably work as well